### ISSUE TITLE
Throw SQLRecoverableException where applicable

### ISSUE THREAD

**Issue by findepi** [prestosql/presto#2645]

**Comment by karthikch53** [comment]
Hi @findepi thanks for looking into the issue. We run queries on a cluster that's shared between apps. There were scenarios when a failed query runs without issues when the app executes the same query moments later. The idea is to look at the error codes so that client can implement the standard retry/backoff logic.
It would be even better if Presto can throw a SQLRecoverableException

**Comment by electrum** [comment]
I don't think this is something that can be solved in only the JDBC driver. The problem is that the concept of recoverable errors is connector specific, but we don't have a way to communicate that today. For example, `HIVE_INVALID_METADATA` is a permanent error, but `HIVE_METASTORE_ERROR` is (likely) a transient error. The engine only knows that these are `EXTERNAL` errors.

**Comment by findepi** [comment]
@electrum than we need a way to access the error codes in JDBC (`SQLException` subclass?)

**Comment by electrum** [comment]
It would be helpful to expose the error name and type. The numeric code is already available via `vendorCode` field of the JDBC driver (which is why we have a numeric code at all). A `SQLException` subclass is one approach. For example, PostgreSQL has `org.postgresql.util.PSQLException` that contains extra error information. Another approach is to provide this as part of the `SQLException` causal chain.

We should survey different drivers and see what they do, and if there are any common or clever approaches.
